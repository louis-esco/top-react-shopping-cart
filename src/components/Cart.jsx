export default function Cart() {
  return <div>This is cart</div>;
}
