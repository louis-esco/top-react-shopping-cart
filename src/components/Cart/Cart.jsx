import { useOutletContext } from "react-router-dom";
import styles from "./Cart.module.css";

export default function Cart() {
  const { getCart, getCartPrice, removeFromCart } = useOutletContext();
  const cart = getCart();
  const cartPrice = getCartPrice().toFixed(2);
  return (
    <>
      <div className="cartList">
        {cart.map((item) => (
          <div key={item.id} className={styles.item}>
            <div className={styles.itemImg}>
              <img src={item.image} alt="" />
            </div>
            <div className="itemContent">
              <div className="itemTitle">{item.title}</div>
              <div className="itemPrice">Unit price: {item.price}€</div>
              <div className="itemQty">Quantity: {item.qty}</div>
              <div className="itemTotal">
                Total price: {item.qty * item.price}€
              </div>
              <button
                onClick={() => {
                  removeFromCart(item, 1);
                }}
              >
                Remove from cart
              </button>
            </div>
          </div>
        ))}
      </div>
      <div className="cartTotal">Cart total price: {cartPrice}€</div>
    </>
  );
}
