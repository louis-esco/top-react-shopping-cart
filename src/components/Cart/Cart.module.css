.emptyCart {
  height: calc(100vh - 200px);
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  gap: 40px;
  color: #3e3e3e;
}

.emptyCartMsg1 {
  font-family: "Segoe Print", "Bradley Hand", Chilanka, TSCu_Comic, casual,
    cursive;
  font-size: 2rem;
}

.cartList {
  padding: 50px 80px;
  display: flex;
  flex-direction: column;
  gap: 25px;
  min-height: calc(100vh - 330px);
}

.cartTotal {
  position: sticky;
  bottom: 0;
  height: 80px;
  background-color: #ffffff;
  padding: 0 40px;
  display: flex;
  justify-content: flex-end;
  align-items: center;
  gap: 30px;
  box-shadow: rgba(0, 0, 0, 0.3) 0px 19px 38px,
    rgba(0, 0, 0, 0.22) 0px 15px 12px;
  border-top: 1px solid #ededed;
  box-sizing: border-box;
}

.priceValue {
  font-weight: 700;
}

.payBtn {
  background-color: #000000;
  color: #ffffff;
  padding: 8px 20px;
  border-radius: 5px;
  border: none;
  font-size: 1.1rem;
}

.payBtn:hover {
  cursor: pointer;
}
