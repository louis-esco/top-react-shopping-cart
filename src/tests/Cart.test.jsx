import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Cart from "../components/Cart/Cart";

const addToCart = vi.fn();
const removeFromCart = vi.fn();
const getCartPrice = () => 156;
const getCartLength = () => 5;
const getCart = () => {
  return [
    {
      id: 1,
      qty: 2,
      title: "Nice item",
      price: 64,
      description: "You can do lots with this item",
      category: "electronics",
      image: "https://fakestoreapi.com/img/61IBBVJvSDL._AC_SY879_.jpg",
      rating: {
        rate: 3.3,
        count: 203,
      },
    },
    {
      id: 2,
      qty: 3,
      title: "Nicer item",
      price: 65,
      description: "You can do lots with this item, and even more",
      category: "electronics",
      image: "https://fakestoreapi.com/img/61IBBVJvSDL._AC_SYUSI879_.jpg",
      rating: {
        rate: 3.5,
        count: 208,
      },
    },
  ];
};

vi.mock("react-router-dom", () => ({
  ...vi.importActual("react-router-dom"),
  useOutletContext: () => ({
    addToCart,
    removeFromCart,
    getCartPrice,
    getCartLength,
    getCart,
  }),
}));

describe("Cart component", () => {
  it("Render Cart", () => {
    const { container } = render(<Cart />);

    expect(container).toMatchSnapshot();
  });

  it("Remove from cart button calls removeFromCart", async () => {
    const user = userEvent.setup();

    render(<Cart />);
    const removeFromCartBtn = screen.getAllByRole("button", {
      name: "âœ• Remove from cart",
    })[0];

    await user.click(removeFromCartBtn);

    expect(removeFromCart).toHaveBeenCalled();
  });
});
