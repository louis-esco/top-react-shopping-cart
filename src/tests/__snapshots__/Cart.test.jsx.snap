// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart component > Render Cart 1`] = `
<div>
  <div
    class="_cartList_f8af68"
  >
    <div
      class="_item_61ad6d"
    >
      <div
        class="_itemImg_61ad6d"
      >
        <img
          alt=""
          src="https://fakestoreapi.com/img/61IBBVJvSDL._AC_SY879_.jpg"
        />
      </div>
      <div
        class="_itemContent_61ad6d"
      >
        <div
          class="_itemTitle_61ad6d"
        >
          Nice item
        </div>
        <div
          class="itemPrice"
        >
          Unit price: 
          64
          €
        </div>
        <div
          class="_itemQty_61ad6d"
        >
          Quantity:
           
          <button
            class="_qtyBtn_61ad6d"
          >
            -
          </button>
           
          <div>
            2
          </div>
           
          <button
            class="_qtyBtn_61ad6d"
          >
            +
          </button>
        </div>
        <div
          class="itemTotal"
        >
          Total price: 
          128
          €
        </div>
        <button
          class="_removeBtn_61ad6d"
        >
          ✕ Remove from cart
        </button>
      </div>
    </div>
    <div
      class="_item_61ad6d"
    >
      <div
        class="_itemImg_61ad6d"
      >
        <img
          alt=""
          src="https://fakestoreapi.com/img/61IBBVJvSDL._AC_SYUSI879_.jpg"
        />
      </div>
      <div
        class="_itemContent_61ad6d"
      >
        <div
          class="_itemTitle_61ad6d"
        >
          Nicer item
        </div>
        <div
          class="itemPrice"
        >
          Unit price: 
          65
          €
        </div>
        <div
          class="_itemQty_61ad6d"
        >
          Quantity:
           
          <button
            class="_qtyBtn_61ad6d"
          >
            -
          </button>
           
          <div>
            3
          </div>
           
          <button
            class="_qtyBtn_61ad6d"
          >
            +
          </button>
        </div>
        <div
          class="itemTotal"
        >
          Total price: 
          195
          €
        </div>
        <button
          class="_removeBtn_61ad6d"
        >
          ✕ Remove from cart
        </button>
      </div>
    </div>
  </div>
  <div
    class="_cartTotal_f8af68"
  >
    <div
      class="_totalPrice_f8af68"
    >
      Cart total price:
       
      <span
        class="_priceValue_f8af68"
      >
        156.00
        €
      </span>
    </div>
    <button
      class="_payBtn_f8af68"
    >
      Checkout & pay
    </button>
  </div>
</div>
`;
